File: ztm-ds-algo/6-binary-trees/q18-count-complete-tree-node/q18-dfs.js
// LeetCode 222.
// https://leetcode.com/problems/count-complete-tree-nodes/

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
 const dfs = function(node, currLevel, height, lastLevelNodes) {
    if(!node) return;
    if(currLevel === height) lastLevelNodes.push(node);
    
    if(node.left) dfs(node.left, currLevel + 1, height, lastLevelNodes);
    if(node.right) dfs(node.right, currLevel + 1, height, lastLevelNodes);
    else return;
}

const calculateHeight = function(root) {
    let temp = root;
    let height = 0;
    while(temp.left) {
        height++;
        temp = temp.left;
    }
    return height;
}

var countNodes = function(root) {
    if(!root) return 0;
    if(root.left === null && root.right === null) return 1;
    
    let count = 0;
    let height = calculateHeight(root);
    count += 2**height - 1;
    
    const lastLevelNodes = []
    dfs(root, 0, height, lastLevelNodes);
    count += lastLevelNodes.length;
    
    return count;
};

// tume: O(n)
// space: O(n)

File: ztm-ds-algo/6-binary-trees/q18-count-complete-tree-node/python/q18-dfs.py
# // LeetCode 222.
# // https://leetcode.com/problems/count-complete-tree-nodes/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def countNodes(self, root):
        def getHeight(root):
            height = 0
            currNode = root
            while currNode.left is not None:
                currNode = currNode.left
                height += 1
            return height
        
        
        def dfs(node, currLevel, lastLevelNum):
            if node is None:
                return
            if currLevel == height:
                lastLevelNum.append(node)
            
            if node.left is not None:
                dfs(node.left, currLevel + 1, lastLevelNum)
            if node.right is not None:
                dfs(node.right, currLevel + 1, lastLevelNum)
            else:
                return
        
        # edge case
        if root is None:
            return 0
        
        # count the node number except the last level 
        height = getHeight(root) # O(logN)
        count = 2**height - 1
        
        # count the last level node number
        lastLevelNum = []
        dfs(root, 0, lastLevelNum)
        
        return count + len(lastLevelNum)

# time: O(n)
# space: O(n)